app/api/items/save_grade.go

```diff
@@ -0,0 +1,275 @@
+		"task_token":   newTaskToken,
+		"bValidated":   validated,
+		"bKeyObtained": keyObtained,
+	})))
+	return service.NoError
+}
+
+func saveGradingResultsIntoDB(store *database.DataStore, user *database.User,
+	requestData *SaveGradeRequest) (validated, keyObtained bool) {
+	const todo = "todo"
+	score := requestData.ScoreToken.Converted.Score
+	userAnswerID := requestData.ScoreToken.Converted.UserAnswerID
+
+	// TODO: handle validation in a proper way (what did he mean??)
+	if score > 99 {
+		validated = true
+	}
+	service.MustNotBeError(store.UserAnswers().ByID(userAnswerID).
+		Where("idUser = ?", user.UserID).
+		Where("idItem = ?", requestData.TaskToken.Converted.LocalItemID).
+		UpdateColumn(map[string]interface{}{
+			"sGradingDate": gorm.Expr("NOW()"),
+			"bValidated":   validated,
+			"iScore":       score,
+		}).Error())
+	// Build query to update users_items
+	// The iScore is set towards the end, so that the IF condition on
+	// sBestAnswerDate is computed before iScore is updated
+	columnsToUpdate := []string{
+		"nbTasksTried",
+		"sLastActivityDate",
+		"sBestAnswerDate",
+		"sLastAnswerDate",
+		"iScore",
+	}
+	values := []interface{}{
+		1,
+		gorm.Expr("NOW()"),
+		gorm.Expr("IF(? > iScore, NOW(), sBestAnswerDate)", score),
+		gorm.Expr("NOW()"),
+		gorm.Expr("GREATEST(?, iScore)", score),
+	}
+	if validated {
+		// Item was validated
+		columnsToUpdate = append(columnsToUpdate,
+			"sAncestorsComputationState", "bValidated", "bKeyObtained", "sValidationDate",
+		)
+		values = append(values,
+			todo, 1, 1, gorm.Expr("IFNULL(sValidationDate, NOW())"))
+		keyObtained = true
+	} else {
+		// Item wasn't validated, check if we unlocked something
+		var unlockedInfo struct {
+			UnlockedItemID string  `gorm:"column:idItemUnlocked"`
+			ScoreMinUnlock float64 `gorm:"column:iScoreMinUnlock"`
+		}
+		service.MustNotBeError(store.Items().ByID(requestData.TaskToken.Converted.LocalItemID).Select("idItemUnlocked, iScoreMinUnlock").
+			Take(&unlockedInfo).Error())
+		if unlockedInfo.UnlockedItemID != "" && unlockedInfo.ScoreMinUnlock < score {
+			keyObtained = true
+			// Update sAncestorsComputationState only if we hadn't obtained the key before
+			columnsToUpdate = append(columnsToUpdate,
+				"sAncestorsComputationState", "bKeyObtained",
+			)
+			values = append(values, gorm.Expr("IF(bKeyObtained = 0, 'todo', sAncestorsComputationState)"), 1)
+		}
+	}
+	if score > 0 && requestData.TaskToken.Converted.AttemptID != nil {
+		// Always propagate attempts if the score was non-zero
+		columnsToUpdate = append(columnsToUpdate, "sAncestorsComputationState")
+		values = append(values, todo)
+	}
+
+	updateExpr := "SET " + strings.Join(columnsToUpdate, " = ?, ") + " = ?"
+	userItemsValues := make([]interface{}, len(values)+2)
+	copy(userItemsValues, values)
+	userItemsValues[len(userItemsValues)-2] = user.UserID
+	userItemsValues[len(userItemsValues)-1] = requestData.TaskToken.Converted.LocalItemID
```

Comment:
Not possible to do these 5lines with an `userItemsValues = append( values, user.UserID, requestData.TaskToken.Converted.LocalItemID)` ?
I assume you would have done it :-)

Reply:
`append(userItemsValues, values..., user.UserID, requestData.TaskToken.Converted.LocalItemID)` is impossible, but is possible to merge the last two lines. Fixed